sdk/protocol2_test.go

```diff
@@ -0,0 +1,171 @@
+		t.Error()
+	}
+	if i.IntegrationVersion != "1.0" {
+		t.Error()
+	}
+	if i.ProtocolVersion != "2" {
+		t.Error()
+	}
+	if len(i.Data) != 0 {
+		t.Error()
+	}
+}
+
+func TestNewIntegrationProtocol2WithDefaultArguments(t *testing.T) {
+	type argumentList struct {
+		sdk_args.DefaultArgumentList
+	}
+
+	var al argumentList
+	i, err := sdk.NewIntegrationProtocol2("TestIntegration", "1.0", &al)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if i.Name != "TestIntegration" {
+		t.Error()
+	}
+	if i.IntegrationVersion != "1.0" {
+		t.Error()
+	}
+	if i.ProtocolVersion != "2" {
+		t.Error()
+	}
+	if len(i.Data) != 0 {
+		t.Error()
+	}
+	if !al.All {
+		t.Error()
+	}
+	if al.Pretty {
+		t.Error()
+	}
+	if al.Verbose {
+		t.Error()
+	}
+}
+
+func TestIntegrationProtocol2_Publish(t *testing.T) {
+	expectedOutput := []byte(`{"name":"TestIntegration","protocol_version":"2","integration_version":"1.0","data":[{"entity":{"name":"EntityOne","type":"test"},"metrics":[{"event_type":"EventTypeForEntityOne","metricOne":99,"metricThree":"test","metricTwo":88}],"inventory":null,"events":null},{"entity":{"name":"EntityTwo","type":"test"},"metrics":[{"event_type":"EventTypeForEntityTwo","metricOne":99,"metricThree":"test","metricTwo":88}],"inventory":null,"events":null}]}`)
+
+	w := testWritter{
+		func(p []byte) {
+			if !bytes.Equal(p, expectedOutput) {
+				t.Errorf("output does not match the expectations.\nExpected:\n%s\nGot:\n%s", expectedOutput, p)
+			}
+		},
+	}
+
+	i, err := sdk.NewIntegrationProtocol2WithWriter("TestIntegration", "1.0", new(struct{}), w)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	e, err := i.Entity("EntityOne", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ms := e.NewMetricSet("EventTypeForEntityOne")
+	ms.SetMetric("metricOne", 99, metric.GAUGE)
+	ms.SetMetric("metricTwo", 88, metric.GAUGE)
+	ms.SetMetric("metricThree", "test", metric.ATTRIBUTE)
+
+	e, err = i.Entity("EntityTwo", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ms = e.NewMetricSet("EventTypeForEntityTwo")
+	ms.SetMetric("metricOne", 99, metric.GAUGE)
+	ms.SetMetric("metricTwo", 88, metric.GAUGE)
+	ms.SetMetric("metricThree", "test", metric.ATTRIBUTE)
+
+	i.Publish()
+}
+
+func TestIntegrationProtocol2_EntityErrorOnEmptyNameAndOrType(t *testing.T) {
+	i, err := sdk.NewIntegrationProtocol2("TestIntegration", "1.0", new(struct{}))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = i.Entity("", "test")
+	if err == nil {
+		t.Fail()
+	}
+
+	_, err = i.Entity("Entity", "")
+	if err == nil {
+		t.Fail()
+	}
+
+	_, err = i.Entity("", "")
+	if err == nil {
+		t.Fail()
+	}
+}
+
+func TestIntegrationProtocol2_EntityReturnsExistingEntity(t *testing.T) {
+	i, err := sdk.NewIntegrationProtocol2("TestIntegration", "1.0", new(struct{}))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	e1, err := i.Entity("Entity1", "test")
+	if err != nil {
+		t.Fail()
+	}
+
+	e2, err := i.Entity("Entity1", "test")
+	if err != nil {
+		t.Fail()
+	}
+
+	if e1 != e2 {
+		t.Error("entity should be equal.")
```

Comment:
I think that this functionality is wrong. According to the requirements:

> Entity would require both a name and a type, and the combination of name+type must be globally unique.

Reply:
Yes, that's why I'm checking that the Entity returned is the SAME we created before.

Reply:
Yeah, you're right! but we should check in the test, that integration contains only one entity, and not 2. At least for me it's not so clear.

Reply:
Ok, make sense! Done :D